File: python-sdks/instrumentations/respan-instrumentation-vertexai/src/respan_instrumentation_vertexai/_translator.py
"""Translate Vertex AI SDK payloads into Respan span fields."""

from __future__ import annotations

import json
from typing import Any, Iterable

from respan_instrumentation_vertexai._constants import (
    ARGS_KEY,
    ASSISTANT_ROLE,
    CANDIDATES_KEY,
    CANDIDATES_TOKEN_COUNT_KEY,
    CONTENT_KEY,
    FUNCTION_CALL_KEY,
    FUNCTION_DECLARATIONS_KEY,
    FUNCTION_KEY,
    FUNCTION_RESPONSE_KEY,
    FUNCTION_TOOL_TYPE,
    GENERATION_CONFIG_KEY,
    ID_KEY,
    MODEL_ROLE,
    NAME_KEY,
    PARAMETERS_KEY,
    PARTS_KEY,
    PROMPT_TOKEN_COUNT_KEY,
    RESPONSE_KEY,
    ROLE_KEY,
    STREAM_KEY,
    SYSTEM_INSTRUCTION_KEY,
    SYSTEM_ROLE,
    TEXT_KEY,
    TOOL_CONFIG_KEY,
    TOOL_ROLE,
    TOOLS_KEY,
    TOTAL_TOKEN_COUNT_KEY,
    TYPE_KEY,
    USAGE_METADATA_KEY,
    USER_ROLE,
)
from respan_sdk.utils.serialization import serialize_value


def safe_json(value: Any) -> str:
    """JSON-encode values after applying the SDK's serializer."""
    try:
        return json.dumps(serialize_value(value=value), default=str)
    except Exception:
        return str(value)


def to_json_attr(value: Any) -> str:
    if isinstance(value, str):
        return value
    return safe_json(value=value)


def _field(value: Any, name: str, default: Any = None) -> Any:
    if isinstance(value, dict):
        return value.get(name, default)
    return getattr(value, name, default)


def _dump_value(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, (str, int, float, bool)):
        return value
    if isinstance(value, dict):
        return {
            str(key): _dump_value(nested_value)
            for key, nested_value in value.items()
            if nested_value is not None
        }
    if isinstance(value, (list, tuple, set)):
        return [_dump_value(item) for item in value if item is not None]
    to_dict = getattr(value, "to_dict", None)
    if callable(to_dict):
        try:
            return to_dict()
        except Exception:
            pass
    model_dump = getattr(value, "model_dump", None)
    if callable(model_dump):
        try:
            return model_dump(exclude_none=True, by_alias=False)
        except TypeError:
            return model_dump()
    return serialize_value(value=value)


def _role(value: Any, default: str = USER_ROLE) -> str:
    role = _field(value, ROLE_KEY, default) or default
    if role == MODEL_ROLE:
        return ASSISTANT_ROLE
    if role == FUNCTION_KEY:
        return TOOL_ROLE
    return str(role)


def _is_content_like(value: Any) -> bool:
    if isinstance(value, dict):
        return PARTS_KEY in value or ROLE_KEY in value
    return hasattr(value, PARTS_KEY) or hasattr(value, ROLE_KEY)


def _is_part_like(value: Any) -> bool:
    if isinstance(value, str):
        return True
    if isinstance(value, dict):
        return any(
            key in value
            for key in (
                TEXT_KEY,
                FUNCTION_CALL_KEY,
                FUNCTION_RESPONSE_KEY,
                "inline_data",
                "file_data",
            )
        )
    return any(
        hasattr(value, attr)
        for attr in (
            TEXT_KEY,
            FUNCTION_CALL_KEY,
            FUNCTION_RESPONSE_KEY,
            "inline_data",
            "file_data",
        )
    )


def _normalize_function_call(value: Any) -> dict[str, Any]:
    function_call: dict[str, Any] = {
        TYPE_KEY: FUNCTION_TOOL_TYPE,
        FUNCTION_KEY: {},
    }
    call_id = _field(value, ID_KEY)
    if call_id:
        function_call[ID_KEY] = call_id
    name = _field(value, NAME_KEY)
    if name:
        function_call[FUNCTION_KEY][NAME_KEY] = name
    args = _field(value, ARGS_KEY)
    if args is not None:
        function_call[FUNCTION_KEY]["arguments"] = to_json_attr(_dump_value(args))
    return function_call


def _normalize_function_response(value: Any) -> dict[str, Any]:
    result: dict[str, Any] = {
        TYPE_KEY: "function_response",
        FUNCTION_KEY: {},
    }
    name = _field(value, NAME_KEY)
    if name:
        result[FUNCTION_KEY][NAME_KEY] = name
    response = _field(value, RESPONSE_KEY)
    if response is not None:
        result[FUNCTION_KEY][RESPONSE_KEY] = _dump_value(response)
    return result


def _normalize_part(part: Any) -> Any:
    if isinstance(part, str):
        return part

    text = _field(part, TEXT_KEY)
    if text is not None:
        return text

    function_call = _field(part, FUNCTION_CALL_KEY)
    if function_call is not None:
        return _normalize_function_call(function_call)

    function_response = _field(part, FUNCTION_RESPONSE_KEY)
    if function_response is not None:
        return _normalize_function_response(function_response)

    dumped = _dump_value(part)
    if isinstance(dumped, dict):
        for key in ("inline_data", "file_data"):
            if dumped.get(key) is not None:
                return {TYPE_KEY: key, key: dumped[key]}
    return dumped


def _normalize_parts(parts: Any) -> Any:
    if parts is None:
        return ""
    if isinstance(parts, (str, bytes)):
        return parts.decode() if isinstance(parts, bytes) else parts
    if not isinstance(parts, (list, tuple)):
        return _normalize_part(parts)

    normalized_parts = [
        normalized
        for part in parts
        if (normalized := _normalize_part(part)) not in (None, "", [], {})
    ]
    if not normalized_parts:
        return ""
    if len(normalized_parts) == 1:
        return normalized_parts[0]
    if all(isinstance(part, str) for part in normalized_parts):
        return "\n".join(normalized_parts)
    return normalized_parts


def _normalize_content(
    content: Any, *, default_role: str = USER_ROLE
) -> dict[str, Any]:
    if isinstance(content, str):
        return {ROLE_KEY: default_role, CONTENT_KEY: content}
    if _is_part_like(content) and not _is_content_like(content):
        return {ROLE_KEY: default_role, CONTENT_KEY: _normalize_part(content)}

    parts = _field(content, PARTS_KEY)
    return {
        ROLE_KEY: _role(content, default=default_role),
        CONTENT_KEY: _normalize_parts(parts),
    }


def normalize_input_messages(
    contents: Any,
    *,
    system_instruction: Any = None,
) -> list[dict[str, Any]]:
    messages: list[dict[str, Any]] = []

    if system_instruction is not None:
        messages.append(
            _normalize_content(system_instruction, default_role=SYSTEM_ROLE)
            if _is_content_like(system_instruction)
            else {
                ROLE_KEY: SYSTEM_ROLE,
                CONTENT_KEY: _normalize_parts(system_instruction),
            }
        )

    if contents is None:
        return messages
    if isinstance(contents, str):
        messages.append({ROLE_KEY: USER_ROLE, CONTENT_KEY: contents})
        return messages
    if _is_content_like(contents):
        messages.append(_normalize_content(contents))
        return messages
    if _is_part_like(contents):
        messages.append({ROLE_KEY: USER_ROLE, CONTENT_KEY: _normalize_part(contents)})
        return messages
    if isinstance(contents, (list, tuple)):
        if all(_is_content_like(item) for item in contents):
            messages.extend(_normalize_content(item) for item in contents)
            return messages
        messages.append({ROLE_KEY: USER_ROLE, CONTENT_KEY: _normalize_parts(contents)})
        return messages

    messages.append({ROLE_KEY: USER_ROLE, CONTENT_KEY: serialize_value(value=contents)})
    return messages


def format_input(contents: Any, *, system_instruction: Any = None) -> str:
    return safe_json(
        value=normalize_input_messages(
            contents=contents,
            system_instruction=system_instruction,
        )
    )


def _candidate_contents(response: Any) -> Iterable[Any]:
    for candidate in _field(response, CANDIDATES_KEY, []) or []:
        content = _field(candidate, CONTENT_KEY)
        if content is not None:
            yield content


def _response_text(response: Any) -> str:
    text = _field(response, TEXT_KEY)
    if isinstance(text, str):
        return text

    text_parts: list[str] = []
    for content in _candidate_contents(response):
        content_value = _normalize_content(content, default_role=ASSISTANT_ROLE).get(
            CONTENT_KEY,
            "",
        )
        if isinstance(content_value, str):
            text_parts.append(content_value)
    return "\n".join(part for part in text_parts if part)


def format_output(response_or_chunks: Any) -> str:
    chunks = response_or_chunks if isinstance(response_or_chunks, list) else None
    if chunks is not None:
        return "".join(_response_text(chunk) for chunk in chunks if chunk is not None)
    if response_or_chunks is None:
        return ""
    return _response_text(response_or_chunks)


def extract_usage(response_or_chunks: Any) -> dict[str, int]:
    response = None
    if isinstance(response_or_chunks, list):
        for chunk in reversed(response_or_chunks):
            if _field(chunk, USAGE_METADATA_KEY) is not None:
                response = chunk
                break
        if response is None and response_or_chunks:
            response = response_or_chunks[-1]
    else:
        response = response_or_chunks

    usage = _field(response, USAGE_METADATA_KEY)
    if usage is None:
        return {}

    result: dict[str, int] = {}
    prompt_tokens = _field(usage, PROMPT_TOKEN_COUNT_KEY)
    completion_tokens = _field(usage, CANDIDATES_TOKEN_COUNT_KEY)
    total_tokens = _field(usage, TOTAL_TOKEN_COUNT_KEY)
    if isinstance(prompt_tokens, int):
        result[PROMPT_TOKEN_COUNT_KEY] = prompt_tokens
    if isinstance(completion_tokens, int):
        result[CANDIDATES_TOKEN_COUNT_KEY] = completion_tokens
    if isinstance(total_tokens, int):
        result[TOTAL_TOKEN_COUNT_KEY] = total_tokens
    return result


def extract_tool_calls(response_or_chunks: Any) -> list[dict[str, Any]]:
    chunks = (
        response_or_chunks
        if isinstance(response_or_chunks, list)
        else [response_or_chunks]
    )
    tool_calls: list[dict[str, Any]] = []
    seen: set[str] = set()
    for response in chunks:
        if response is None:
            continue
        for content in _candidate_contents(response):
            parts = _field(content, PARTS_KEY, []) or []
            for part in parts:
                function_call = _field(part, FUNCTION_CALL_KEY)
                if function_call is None:
                    continue
                normalized = _normalize_function_call(function_call)
                function = normalized.get(FUNCTION_KEY)
                if not isinstance(function, dict) or not function.get(NAME_KEY):
                    continue
                signature = safe_json(value=normalized)
                if signature in seen:
                    continue
                seen.add(signature)
                tool_calls.append(normalized)
    return tool_calls


def _function_declaration_tool(definition: Any) -> dict[str, Any] | None:
    name = _field(definition, NAME_KEY)
    if not name:
        return None
    function: dict[str, Any] = {NAME_KEY: name}
    description = _field(definition, "description")
    if description:
        function["description"] = description
    parameters = _field(definition, PARAMETERS_KEY)
    if parameters is not None:
        function[PARAMETERS_KEY] = _dump_value(parameters)
    return {TYPE_KEY: FUNCTION_TOOL_TYPE, FUNCTION_KEY: function}


def extract_tools(tools: Any) -> list[dict[str, Any]]:
    if not tools:
        return []

    normalized_tools: list[dict[str, Any]] = []
    for tool in tools:
        declarations = _field(tool, FUNCTION_DECLARATIONS_KEY) or []
        if isinstance(tool, dict) and FUNCTION_DECLARATIONS_KEY not in tool:
            normalized = _function_declaration_tool(tool)
            if normalized is not None:
                normalized_tools.append(normalized)
        for declaration in declarations:
            normalized = _function_declaration_tool(declaration)
            if normalized is not None:
                normalized_tools.append(normalized)

    return normalized_tools


def _first_arg(args: tuple[Any, ...]) -> Any:
    return args[0] if args else None


def _model_name(instance: Any) -> str | None:
    for attr_name in ("_model_name", "model_name", "_model_id", "model_id"):
        value = getattr(instance, attr_name, None)
        if value:
            return str(value)
    nested_model = getattr(instance, "model", None)
    if nested_model is not None and nested_model is not instance:
        return _model_name(nested_model)
    return None


def _system_instruction(instance: Any) -> Any:
    for attr_name in (SYSTEM_INSTRUCTION_KEY, f"_{SYSTEM_INSTRUCTION_KEY}"):
        value = getattr(instance, attr_name, None)
        if value is not None:
            return value
    nested_model = getattr(instance, "model", None)
    if nested_model is not None and nested_model is not instance:
        return _system_instruction(nested_model)
    return None


def _tools(instance: Any, kwargs: dict[str, Any]) -> Any:
    if kwargs.get(TOOLS_KEY) is not None:
        return kwargs.get(TOOLS_KEY)
    for attr_name in (TOOLS_KEY, f"_{TOOLS_KEY}"):
        value = getattr(instance, attr_name, None)
        if value is not None:
            return value
    nested_model = getattr(instance, "model", None)
    if nested_model is not None and nested_model is not instance:
        return _tools(nested_model, kwargs)
    return None


def request_payload_from_call(
    *,
    instance: Any,
    args: tuple[Any, ...],
    kwargs: dict[str, Any],
) -> dict[str, Any]:
    contents = kwargs.get("contents", kwargs.get("content", _first_arg(args)))
    return {
        "model": _model_name(instance),
        "contents": contents,
        GENERATION_CONFIG_KEY: kwargs.get(GENERATION_CONFIG_KEY),
        TOOL_CONFIG_KEY: kwargs.get(TOOL_CONFIG_KEY),
        TOOLS_KEY: _tools(instance, kwargs),
        STREAM_KEY: bool(kwargs.get(STREAM_KEY, False)),
        SYSTEM_INSTRUCTION_KEY: _system_instruction(instance),
    }

File: javascript-sdks/instrumentations/respan-instrumentation-vertexai/src/_translator.ts
import {
  ARGS_KEY,
  ASSISTANT_ROLE,
  CANDIDATES_KEY,
  CANDIDATES_TOKEN_COUNT_KEY,
  CANDIDATES_TOKEN_COUNT_SNAKE_KEY,
  THOUGHTS_TOKEN_COUNT_KEY,
  THOUGHTS_TOKEN_COUNT_SNAKE_KEY,
  CONTENT_KEY,
  FUNCTION_CALL_KEY,
  FUNCTION_CALL_SNAKE_KEY,
  FUNCTION_DECLARATIONS_KEY,
  FUNCTION_DECLARATIONS_SNAKE_KEY,
  FUNCTION_KEY,
  FUNCTION_RESPONSE_KEY,
  FUNCTION_RESPONSE_SNAKE_KEY,
  FUNCTION_TOOL_TYPE,
  GENERATION_CONFIG_KEY,
  GENERATION_CONFIG_SNAKE_KEY,
  ID_KEY,
  MODEL_ROLE,
  NAME_KEY,
  PARAMETERS_KEY,
  PARTS_KEY,
  PROMPT_TOKEN_COUNT_KEY,
  PROMPT_TOKEN_COUNT_SNAKE_KEY,
  RESPONSE_KEY,
  ROLE_KEY,
  STREAM_KEY,
  SYSTEM_INSTRUCTION_KEY,
  SYSTEM_INSTRUCTION_SNAKE_KEY,
  SYSTEM_ROLE,
  TEXT_KEY,
  TOOL_CONFIG_KEY,
  TOOL_CONFIG_SNAKE_KEY,
  TOOL_ROLE,
  TOOLS_KEY,
  TOTAL_TOKEN_COUNT_KEY,
  TOTAL_TOKEN_COUNT_SNAKE_KEY,
  TYPE_KEY,
  USER_ROLE,
  USAGE_METADATA_KEY,
  USAGE_METADATA_SNAKE_KEY,
} from "./_constants.js";

export interface VertexAIRequestPayload {
  model?: string;
  contents?: unknown;
  generationConfig?: unknown;
  toolConfig?: unknown;
  tools?: unknown;
  stream?: boolean;
  systemInstruction?: unknown;
}

export function safeJson(value: unknown): string {
  try {
    return JSON.stringify(value, (_key, innerValue) =>
      typeof innerValue === "bigint" ? innerValue.toString() : innerValue,
    );
  } catch {
    return String(value);
  }
}

export function toSerializableValue(value: any): any {
  if (value === null) return null;
  if (value === undefined) return undefined;
  if (
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean"
  ) {
    return value;
  }
  if (typeof value === "bigint") {
    return value.toString();
  }
  if (value instanceof Date) {
    return value.toISOString();
  }
  if (Array.isArray(value)) {
    return value.map((item) => toSerializableValue(item));
  }
  if (typeof value === "object") {
    if (typeof value.toJSON === "function") {
      try {
        return toSerializableValue(value.toJSON());
      } catch {
        // Continue with a structural copy.
      }
    }
    if (typeof value.toObject === "function") {
      try {
        return toSerializableValue(value.toObject());
      } catch {
        // Continue with a structural copy.
      }
    }

    const normalized: Record<string, unknown> = {};
    for (const [key, itemValue] of Object.entries(value as Record<string, unknown>)) {
      if (itemValue !== undefined) {
        normalized[key] = toSerializableValue(itemValue);
      }
    }
    return normalized;
  }
  return String(value);
}

export function toJsonAttr(value: unknown): string {
  if (typeof value === "string") return value;
  return safeJson(toSerializableValue(value));
}

function getField(value: any, ...names: string[]): any {
  if (value === undefined || value === null) return undefined;
  for (const name of names) {
    if (typeof value === "object" && name in value) {
      return value[name];
    }
    const nestedValue = value?.[name];
    if (nestedValue !== undefined) {
      return nestedValue;
    }
  }
  return undefined;
}

function firstDefined<T>(...values: Array<T | undefined>): T | undefined {
  for (const value of values) {
    if (value !== undefined) return value;
  }
  return undefined;
}

function role(value: unknown, defaultRole = USER_ROLE): string {
  const rawRole = getField(value, ROLE_KEY) ?? defaultRole;
  if (rawRole === MODEL_ROLE) return ASSISTANT_ROLE;
  if (rawRole === FUNCTION_KEY) return TOOL_ROLE;
  return String(rawRole);
}

function isContentLike(value: unknown): boolean {
  return Boolean(value && typeof value === "object" && (
    getField(value, PARTS_KEY) !== undefined ||
    getField(value, ROLE_KEY) !== undefined
  ));
}

function isPartLike(value: unknown): boolean {
  if (typeof value === "string") return true;
  return Boolean(value && typeof value === "object" && (
    getField(value, TEXT_KEY) !== undefined ||
    getField(value, FUNCTION_CALL_KEY, FUNCTION_CALL_SNAKE_KEY) !== undefined ||
    getField(value, FUNCTION_RESPONSE_KEY, FUNCTION_RESPONSE_SNAKE_KEY) !== undefined ||
    getField(value, "inlineData", "inline_data") !== undefined ||
    getField(value, "fileData", "file_data") !== undefined
  ));
}

function normalizeFunctionCall(value: any): Record<string, unknown> {
  const functionCall: Record<string, any> = {
    [TYPE_KEY]: FUNCTION_TOOL_TYPE,
    [FUNCTION_KEY]: {},
  };
  const callId = getField(value, ID_KEY);
  if (callId) functionCall[ID_KEY] = callId;

  const name = getField(value, NAME_KEY);
  if (name) functionCall[FUNCTION_KEY][NAME_KEY] = name;

  const args = getField(value, ARGS_KEY);
  if (args !== undefined) {
    functionCall[FUNCTION_KEY].arguments = toJsonAttr(toSerializableValue(args));
  }
  return functionCall;
}

function normalizeFunctionResponse(value: any): Record<string, unknown> {
  const result: Record<string, any> = {
    [TYPE_KEY]: "function_response",
    [FUNCTION_KEY]: {},
  };
  const name = getField(value, NAME_KEY);
  if (name) result[FUNCTION_KEY][NAME_KEY] = name;

  const response = getField(value, RESPONSE_KEY);
  if (response !== undefined) {
    result[FUNCTION_KEY][RESPONSE_KEY] = toSerializableValue(response);
  }
  return result;
}

function normalizePart(part: unknown): unknown {
  if (typeof part === "string") return part;

  const text = getField(part, TEXT_KEY);
  if (text !== undefined) return text;

  const functionCall = getField(part, FUNCTION_CALL_KEY, FUNCTION_CALL_SNAKE_KEY);
  if (functionCall !== undefined) {
    return normalizeFunctionCall(functionCall);
  }

  const functionResponse = getField(part, FUNCTION_RESPONSE_KEY, FUNCTION_RESPONSE_SNAKE_KEY);
  if (functionResponse !== undefined) {
    return normalizeFunctionResponse(functionResponse);
  }

  const inlineData = getField(part, "inlineData", "inline_data");
  if (inlineData !== undefined) {
    return { [TYPE_KEY]: "inline_data", inline_data: toSerializableValue(inlineData) };
  }

  const fileData = getField(part, "fileData", "file_data");
  if (fileData !== undefined) {
    return { [TYPE_KEY]: "file_data", file_data: toSerializableValue(fileData) };
  }

  return toSerializableValue(part);
}

function normalizeParts(parts: unknown): unknown {
  if (parts === undefined || parts === null) return "";
  if (typeof parts === "string") return parts;
  if (!Array.isArray(parts)) return normalizePart(parts);

  const normalizedParts = parts
    .map((part) => normalizePart(part))
    .filter((part) => part !== undefined && part !== null && part !== "");
  if (normalizedParts.length === 0) return "";
  if (normalizedParts.length === 1) return normalizedParts[0];
  if (normalizedParts.every((part) => typeof part === "string")) {
    return normalizedParts.join("\n");
  }
  return normalizedParts;
}

function normalizeContent(content: unknown, defaultRole = USER_ROLE): Record<string, unknown> {
  if (typeof content === "string") {
    return { [ROLE_KEY]: defaultRole, [CONTENT_KEY]: content };
  }
  if (isPartLike(content) && !isContentLike(content)) {
    return { [ROLE_KEY]: defaultRole, [CONTENT_KEY]: normalizePart(content) };
  }

  return {
    [ROLE_KEY]: role(content, defaultRole),
    [CONTENT_KEY]: normalizeParts(getField(content, PARTS_KEY)),
  };
}

export function normalizeInputMessages(
  contents: unknown,
  systemInstruction?: unknown,
): Array<Record<string, unknown>> {
  const messages: Array<Record<string, unknown>> = [];

  if (systemInstruction !== undefined && systemInstruction !== null) {
    messages.push(
      isContentLike(systemInstruction)
        ? normalizeContent(systemInstruction, SYSTEM_ROLE)
        : { [ROLE_KEY]: SYSTEM_ROLE, [CONTENT_KEY]: normalizeParts(systemInstruction) },
    );
  }

  if (contents === undefined || contents === null) return messages;
  if (typeof contents === "string") {
    messages.push({ [ROLE_KEY]: USER_ROLE, [CONTENT_KEY]: contents });
    return messages;
  }
  if (isContentLike(contents)) {
    messages.push(normalizeContent(contents));
    return messages;
  }
  if (isPartLike(contents)) {
    messages.push({ [ROLE_KEY]: USER_ROLE, [CONTENT_KEY]: normalizePart(contents) });
    return messages;
  }
  if (Array.isArray(contents)) {
    if (contents.every((item) => isContentLike(item))) {
      messages.push(...contents.map((item) => normalizeContent(item)));
      return messages;
    }
    messages.push({ [ROLE_KEY]: USER_ROLE, [CONTENT_KEY]: normalizeParts(contents) });
    return messages;
  }

  messages.push({ [ROLE_KEY]: USER_ROLE, [CONTENT_KEY]: toSerializableValue(contents) });
  return messages;
}

export function formatInput(contents: unknown, systemInstruction?: unknown): string {
  return safeJson(normalizeInputMessages(contents, systemInstruction));
}

function candidateContents(response: unknown): unknown[] {
  const candidates = getField(response, CANDIDATES_KEY) ?? [];
  if (!Array.isArray(candidates)) return [];
  return candidates
    .map((candidate) => getField(candidate, CONTENT_KEY))
    .filter((content) => content !== undefined && content !== null);
}

function responseText(response: unknown): string {
  if (response === undefined || response === null) return "";

  const textProp = getField(response, TEXT_KEY);
  if (typeof textProp === "string") return textProp;
  if (typeof textProp === "function") {
    try {
      const text = textProp.call(response);
      if (typeof text === "string") return text;
    } catch {
      // Fall back to candidates.
    }
  }

  const parts: string[] = [];
  for (const content of candidateContents(response)) {
    const normalized = normalizeContent(content, ASSISTANT_ROLE)[CONTENT_KEY];
    if (typeof normalized === "string") {
      parts.push(normalized);
    }
  }
  return parts.join("\n");
}

export function formatOutput(responseOrChunks: unknown): string {
  if (Array.isArray(responseOrChunks)) {
    return responseOrChunks.map((chunk) => responseText(chunk)).join("");
  }
  return responseText(responseOrChunks);
}

export function extractUsage(responseOrChunks: unknown): Record<string, number> {
  let response = responseOrChunks;
  if (Array.isArray(responseOrChunks)) {
    response = [...responseOrChunks]
      .reverse()
      .find((chunk) => getField(chunk, USAGE_METADATA_KEY, USAGE_METADATA_SNAKE_KEY) !== undefined)
      ?? responseOrChunks.at(-1);
  }

  const usage = getField(response, USAGE_METADATA_KEY, USAGE_METADATA_SNAKE_KEY);
  if (!usage) return {};

  const result: Record<string, number> = {};
  const promptTokens = getField(usage, PROMPT_TOKEN_COUNT_KEY, PROMPT_TOKEN_COUNT_SNAKE_KEY);
  const completionTokens = getField(
    usage,
    CANDIDATES_TOKEN_COUNT_KEY,
    CANDIDATES_TOKEN_COUNT_SNAKE_KEY,
  );
  const thoughtsTokens = getField(
    usage,
    THOUGHTS_TOKEN_COUNT_KEY,
    THOUGHTS_TOKEN_COUNT_SNAKE_KEY,
  );
  const totalTokens = getField(usage, TOTAL_TOKEN_COUNT_KEY, TOTAL_TOKEN_COUNT_SNAKE_KEY);

  if (typeof promptTokens === "number") result[PROMPT_TOKEN_COUNT_KEY] = promptTokens;
  if (typeof completionTokens === "number") {
    // Gemini reports thinking tokens separately from candidatesTokenCount and bills
    // them at the output rate, so they belong in the output count. This matches the
    // google-adk instrumentation, which already folds them in.
    result[CANDIDATES_TOKEN_COUNT_KEY] =
      completionTokens + (typeof thoughtsTokens === "number" ? thoughtsTokens : 0);
  }
  if (typeof totalTokens === "number") result[TOTAL_TOKEN_COUNT_KEY] = totalTokens;
  return result;
}

export function extractToolCalls(responseOrChunks: unknown): Array<Record<string, unknown>> {
  const chunks = Array.isArray(responseOrChunks) ? responseOrChunks : [responseOrChunks];
  const toolCalls: Array<Record<string, unknown>> = [];
  const seen = new Set<string>();

  for (const response of chunks) {
    for (const content of candidateContents(response)) {
      const parts = getField(content, PARTS_KEY) ?? [];
      if (!Array.isArray(parts)) continue;
      for (const part of parts) {
        const functionCall = getField(part, FUNCTION_CALL_KEY, FUNCTION_CALL_SNAKE_KEY);
        if (functionCall === undefined || functionCall === null) continue;
        const normalized = normalizeFunctionCall(functionCall);
        const functionValue = normalized[FUNCTION_KEY];
        if (
          !functionValue ||
          typeof functionValue !== "object" ||
          !(functionValue as Record<string, unknown>)[NAME_KEY]
        ) {
          continue;
        }
        const signature = safeJson(normalized);
        if (seen.has(signature)) continue;
        seen.add(signature);
        toolCalls.push(normalized);
      }
    }
  }

  return toolCalls;
}

function functionDeclarationTool(definition: unknown): Record<string, unknown> | undefined {
  const name = getField(definition, NAME_KEY);
  if (!name) return undefined;

  const functionValue: Record<string, unknown> = { [NAME_KEY]: name };
  const description = getField(definition, "description");
  if (description) functionValue.description = description;

  const parameters = getField(definition, PARAMETERS_KEY);
  if (parameters !== undefined) {
    functionValue[PARAMETERS_KEY] = toSerializableValue(parameters);
  }

  return { [TYPE_KEY]: FUNCTION_TOOL_TYPE, [FUNCTION_KEY]: functionValue };
}

export function extractTools(tools: unknown): Array<Record<string, unknown>> {
  if (!Array.isArray(tools)) return [];

  const normalizedTools: Array<Record<string, unknown>> = [];
  for (const tool of tools) {
    const declarations = getField(
      tool,
      FUNCTION_DECLARATIONS_KEY,
      FUNCTION_DECLARATIONS_SNAKE_KEY,
    );
    if (Array.isArray(declarations)) {
      for (const declaration of declarations) {
        const normalized = functionDeclarationTool(declaration);
        if (normalized) normalizedTools.push(normalized);
      }
      continue;
    }

    const normalized = functionDeclarationTool(tool);
    if (normalized) normalizedTools.push(normalized);
  }

  return normalizedTools;
}

function firstArg(args: unknown[]): unknown {
  return args.length > 0 ? args[0] : undefined;
}

function requestObjectFromArgs(args: unknown[]): Record<string, any> {
  const value = firstArg(args);
  return value && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, any>
    : {};
}

function modelName(instance: any): string | undefined {
  const value = firstDefined(
    getField(instance, "model"),
    getField(instance, "modelName", "model_name"),
    getField(instance, "_model", "_modelName", "_model_name"),
    getField(instance, "modelId", "model_id", "_modelId", "_model_id"),
  );
  if (typeof value === "string" && value.length > 0) return value;
  if (value && typeof value === "object" && value !== instance) {
    const nested = modelName(value);
    if (nested) return nested;
  }

  const nestedModel = getField(instance, "generativeModel", "modelInstance");
  if (nestedModel && nestedModel !== instance) return modelName(nestedModel);
  return undefined;
}

function systemInstruction(instance: any, request: Record<string, any>): unknown {
  const requestValue = getField(request, SYSTEM_INSTRUCTION_KEY, SYSTEM_INSTRUCTION_SNAKE_KEY);
  if (requestValue !== undefined) return requestValue;

  const instanceValue = getField(
    instance,
    SYSTEM_INSTRUCTION_KEY,
    SYSTEM_INSTRUCTION_SNAKE_KEY,
    "_systemInstruction",
    "_system_instruction",
  );
  if (instanceValue !== undefined) return instanceValue;

  const nestedModel = getField(instance, "model", "generativeModel", "modelInstance");
  if (nestedModel && nestedModel !== instance) {
    return systemInstruction(nestedModel, request);
  }
  return undefined;
}

function tools(instance: any, request: Record<string, any>): unknown {
  const requestTools = getField(request, TOOLS_KEY);
  if (requestTools !== undefined) return requestTools;

  const instanceTools = getField(instance, TOOLS_KEY, "_tools");
  if (instanceTools !== undefined) return instanceTools;

  const nestedModel = getField(instance, "model", "generativeModel", "modelInstance");
  if (nestedModel && nestedModel !== instance) {
    return tools(nestedModel, request);
  }
  return undefined;
}

function generationConfig(instance: any, request: Record<string, any>): unknown {
  return firstDefined(
    getField(request, GENERATION_CONFIG_KEY, GENERATION_CONFIG_SNAKE_KEY),
    getField(instance, GENERATION_CONFIG_KEY, GENERATION_CONFIG_SNAKE_KEY),
  );
}

export function requestPayloadFromCall(
  instance: unknown,
  args: unknown[],
  opts: { isChatMethod?: boolean; isStreamMethod?: boolean } = {},
): VertexAIRequestPayload {
  const request = requestObjectFromArgs(args);
  const rawFirstArg = firstArg(args);
  const contents = opts.isChatMethod
    ? rawFirstArg
    : firstDefined(getField(request, "contents", CONTENT_KEY), rawFirstArg);

  return {
    model: modelName(instance),
    contents,
    generationConfig: generationConfig(instance, request),
    toolConfig: getField(request, TOOL_CONFIG_KEY, TOOL_CONFIG_SNAKE_KEY),
    tools: tools(instance, request),
    stream: opts.isStreamMethod || Boolean(getField(request, STREAM_KEY)),
    systemInstruction: systemInstruction(instance, request),
  };
}
